File: src/main/java/org/apache/sysml/hops/spoof2/rewrite/SNodeRewriteUtils.java

package org.apache.sysml.hops.spoof2.rewrite;

import java.util.ArrayList;
import org.apache.sysml.hops.spoof2.plan.SNode;

public class SNodeRewriteUtils 
{
	public static int getChildReferencePos( SNode parent, SNode child ) {
		return parent.getInput().indexOf(child);
	}
	
	public static void removeChildReference( SNode parent, SNode child ) {
		parent.getInput().remove( child );
		child.getParent().remove( parent );
	}
	
	public static void removeChildReferenceByPos( SNode parent, SNode child, int posChild ) {
		parent.getInput().remove( posChild );
		child.getParent().remove( parent );
	}

	public static void removeAllChildReferences( SNode parent )
	{
		//remove parent reference from all childs
		for( SNode child : parent.getInput() )
			child.getParent().remove(parent);
		
		//remove all child references
		parent.getInput().clear();
	}
	
	public static void addChildReference( SNode parent, SNode child ) {
		parent.getInput().add( child );
		child.getParent().add( parent );
	}
	
	public static void addChildReference( SNode parent, SNode child, int pos ){
		parent.getInput().add( pos, child );
		child.getParent().add( parent );
	}
	
	public static void rewireAllParentChildReferences( SNode hold, SNode hnew ) {
		ArrayList<SNode> parents = new ArrayList<SNode>(hold.getParent());
		for( SNode lparent : parents )
			SNodeRewriteUtils.replaceChildReference(lparent, hold, hnew);
	}
	
	public static void replaceChildReference( SNode parent, SNode inOld, SNode inNew ) {
		int pos = getChildReferencePos(parent, inOld);
		removeChildReferenceByPos(parent, inOld, pos);
		addChildReference(parent, inNew, pos);
	}
	
	public static void replaceChildReference( SNode parent, SNode inOld, SNode inNew, int pos ) {
		replaceChildReference(parent, inOld, inNew, pos, true);
	}
	
	public static void replaceChildReference( SNode parent, SNode inOld, SNode inNew, int pos, boolean refresh ) {
		removeChildReferenceByPos(parent, inOld, pos);
		addChildReference(parent, inNew, pos);
	}
	
	public static void cleanupUnreferenced( SNode... inputs ) {
		for( SNode input : inputs )
			if( input.getParent().isEmpty() )
				removeAllChildReferences(input);
	}
}

File: src/main/java/org/apache/sysml/hops/spoof2/rewrite/SNodeRewriteUtils.kt
package org.apache.sysml.hops.spoof2.rewrite

import java.util.ArrayList
import org.apache.sysml.hops.spoof2.plan.SNode
import org.apache.sysml.hops.spoof2.plan.SNodeNary
import org.apache.sysml.hops.spoof2.plan.SNodeNary.NaryOp

object SNodeRewriteUtils {
    fun getChildReferencePos(parent: SNode, child: SNode): Int {
        return parent.inputs.indexOf(child)
    }

    fun removeChildReference(parent: SNode, child: SNode) {
        parent.inputs.remove(child)
        child.parents.remove(parent)
    }

    fun removeChildReferenceByPos(parent: SNode, child: SNode, posChild: Int) {
        parent.inputs.removeAt(posChild)
        child.parents.remove(parent)
    }

    fun removeAllChildReferences(parent: SNode) {
        //remove parent reference from all childs
        for (child in parent.inputs)
            child.parents.remove(parent)

        //remove all child references
        parent.inputs.clear()
    }

    fun addChildReference(parent: SNode, child: SNode) {
        parent.inputs.add(child)
        child.parents.add(parent)
    }

    fun addChildReference(parent: SNode, child: SNode, pos: Int) {
        parent.inputs.add(pos, child)
        child.parents.add(parent)
    }

    fun rewireAllParentChildReferences(hold: SNode, hnew: SNode) {
        val parents = ArrayList(hold.parents)
        for (lparent in parents)
            SNodeRewriteUtils.replaceChildReference(lparent, hold, hnew)
    }

    fun replaceChildReference(parent: SNode, inOld: SNode, inNew: SNode) {
        val pos = getChildReferencePos(parent, inOld)
        removeChildReferenceByPos(parent, inOld, pos)
        addChildReference(parent, inNew, pos)
    }

    @JvmOverloads fun replaceChildReference(parent: SNode, inOld: SNode, inNew: SNode, pos: Int, refresh: Boolean = true) {
        removeChildReferenceByPos(parent, inOld, pos)
        addChildReference(parent, inNew, pos)
    }

    fun cleanupUnreferenced(vararg inputs: SNode) {
        for (input in inputs)
            if (input.parents.isEmpty())
                removeAllChildReferences(input)
    }

    fun isSNodeNary(node: SNode, type: NaryOp): Boolean {
        return node is SNodeNary && node.op === type
    }
}
